acceptance_testing.go

```diff
@@ -0,0 +1,834 @@
+		`(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?`)
+	is.True(semverRegex.MatchString(spec.Version)) // Specification.Version is not a valid semantic version (vX.Y.Z)
+}
+
+func (a acceptanceTest) TestSource_Configure_Success(t *testing.T) {
+	a.skipIfNoSource(t)
+	is := is.New(t)
+	ctx := context.Background()
+	defer goleak.VerifyNone(t, a.driver.GoleakOptions(t)...)
+
+	source := a.driver.Connector().NewSource()
+	err := source.Configure(ctx, a.driver.SourceConfig(t))
+	is.NoErr(err)
+
+	// calling Teardown after Configure is valid and happens when connector is created
+	err = source.Teardown(ctx)
+	is.NoErr(err)
+}
+
+func (a acceptanceTest) TestSource_Configure_RequiredParams(t *testing.T) {
+	a.skipIfNoSpecification(t)
+	a.skipIfNoSource(t)
+	is := is.New(t)
+	ctx := context.Background()
+
+	spec := a.driver.Connector().NewSpecification()
+	origCfg := a.driver.SourceConfig(t)
+
+	for name, p := range spec.SourceParams {
+		if p.Required {
+			// removing the required parameter from the config should provoke an error
+			t.Run(name, func(t *testing.T) {
+				haveCfg := a.cloneConfig(origCfg)
+				delete(haveCfg, name)
+
+				is.Equal(len(haveCfg)+1, len(origCfg)) // source config does not contain required parameter, please check the test setup
+
+				source := a.driver.Connector().NewSource()
+				err := source.Configure(ctx, haveCfg)
+				is.True(err != nil)
+
+				err = source.Teardown(ctx)
+				is.NoErr(err)
+			})
+		}
+	}
+}
+
+func (a acceptanceTest) TestSource_Open_ResumeAtPosition(t *testing.T) {
+	a.skipIfNoSource(t)
+	is := is.New(t)
+	ctx := context.Background()
+	defer goleak.VerifyNone(t, a.driver.GoleakOptions(t)...)
+
+	source, sourceCleanup := a.openSource(ctx, t, nil) // listen from beginning
+	defer sourceCleanup()
+
+	// write expectations
+	want := a.generateRecords(10)
+	a.driver.Write(t, &want)
+
+	// read all records, but stop acking them after we read half of them
+	var lastPosition Position
+	var lastIndex int
+	for i := 0; i < len(want); i++ {
+		readCtx, cancel := context.WithTimeout(ctx, time.Second*5)
+		defer cancel()
+
+		got, err := source.Read(readCtx)
+		is.NoErr(err)
+
+		if i < len(want)/2 {
+			err = source.Ack(ctx, got.Position)
+			is.NoErr(err)
+			lastPosition = got.Position
+			lastIndex = i
+		}
+	}
+	sourceCleanup()
+
+	// At this point we have read all records but acked only half and stopped
+	// the connector. We should be able to resume from the last record that was
+	// not acked.
+
+	source, sourceCleanup = a.openSource(ctx, t, lastPosition) // listen from position
+	defer sourceCleanup()
+
+	for i := lastIndex + 1; i < len(want); i++ {
+		readCtx, cancel := context.WithTimeout(ctx, time.Second*5)
+		defer cancel()
+
+		got, err := source.Read(readCtx)
+		is.NoErr(err)
+
+		err = source.Ack(ctx, got.Position)
+		is.NoErr(err)
+
+		want[i].Position = got.Position   // position can't be determined in advance
+		want[i].CreatedAt = got.CreatedAt // created at can't be determined in advance
```

Comment:
So this implicitly also asserts that records are read in the same order in which they were originally written to the source system?

Reply:
I'm wondering if it's too strong of a requirement or if certain connectors will ever meet it. E.g. the S3 connector: from what I recall, we have no way of knowing the insertion order. When listing S3 objects, they are sorted by key.

Reply:
Good point 🤔 IMO it would feel wrong to just accept any order though, it feels too vague if we let the source produce just a collection of records in any order. I'm using the word "feels", I have no hard data to back that up 🙃 Other than maybe records could reference each other, so order is important in that case (e.g. record B references record A, so ideally A is produced before B).

I thought about if there's a middle ground. We could say the connector should be deterministic, in the sense that it always produces the same order. It could be _any_ order, as long as it's always the same order. This is already achievable by writing your own driver and changing the order of the records in the expectation:
```
type AcceptanceTestDriver struct {
	sdk.ConfigurableAcceptanceTestDriver
}

func (d AcceptanceTestDriver) WriteToSource(t *testing.T, records *[]sdk.Record) {
	// sort records by key ...
	d.ConfigurableAcceptanceTestDriver.WriteToSource(t, records)
}
```

Thoughts?

Reply:
I just noticed that it's not possible to do the same in `ReadFromDestination`, we would have to pass in the slice of the expectations so it could be changed.

Reply:
Alright, so I changed the signature of `ReadFromDestination` and `WriteToSource` so the parameter is a slice of records that should be read/written and the returned slice is what actually was read/written. This means that the order of the records can be changed as the driver sees fit.

Reply:
Sounds good! To the order: it _feels_ (there it is again), that the order is important in a minority of cases, and it's probably limited to those cases where there is some kind of relationship between records. When it comes to at least some sources I've worked with (RDBMs, Kafka) you can't always guarantee the ordering, so maybe a default of no ordering assumed in tests works best.

Reply:
The order of records shouldn't play a role anymore now!

Reply:
> The order of records shouldn't play a role anymore now!

Anarchy loading... :D